File: templates/src/utils/PointIcons.ts
/**
 * PointIcons.ts
 * Mapa de chaves FA (kebab-case) para IconDefinition e funções de renderização
 * de pino SVG para uso com ol/style/Icon.
 */
import type { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import {
    faLocationDot, faMapPin, faMapMarkerAlt,
    faStar, faFlag, faCrosshairs, faCompass, faCircle, faCircleDot, faSignsPost,
    faCar, faBus, faBicycle, faMotorcycle, faTruck, faTrain, faPlane, faHelicopter,
    faShip, faTaxi, faGasPump, faParking, faTrafficLight, faRoad, faRoute,
    faTruckMedical, faPersonBiking, faPersonHiking, faPlug,
    faHospital, faHouseMedical, faPills, faHeartPulse, faSyringe, faStethoscope,
    faWheelchair, faCross, faKitMedical, faPersonCane,
    faSchool, faBuildingColumns, faGraduationCap, faChalkboardUser, faMicroscope,
    faBookOpen, faPencil, faBook,
    faUtensils, faMugHot, faStore, faCartShopping, faBed, faLandmark, faEnvelope,
    faScissors, faWrench, faCity, faBuilding, faIndustry, faWarehouse,
    faShieldHalved, faFireExtinguisher, faLifeRing, faTriangleExclamation,
    faBell, faPersonShelter, faVault,
    faFutbol, faDumbbell, faMasksTheater, faCamera, faPersonSwimming, faMountainSun,
    faTree, faMusic, faPanorama,
    faChurch, faPlaceOfWorship, faChessRook,
    faBolt, faDroplet, faTowerBroadcast, faSolarPanel, faRecycle, faWater,
    faLeaf, faFire, faSatellite,
    faCircleExclamation, faCircleInfo, faCircleQuestion, faBan, faCircleCheck,
    faArrowUp, faArrowRight,
} from '@fortawesome/free-solid-svg-icons';

// ─────────────────────────────────────────────────────────────────────────────
// Mapa completo: kebab-key → IconDefinition
// ─────────────────────────────────────────────────────────────────────────────
export const POINT_ICONS: Record<string, IconDefinition> = {
    'location-dot': faLocationDot,
    'map-pin': faMapPin,
    'map-marker-alt': faMapMarkerAlt,
    'star': faStar,
    'flag': faFlag,
    'crosshairs': faCrosshairs,
    'compass': faCompass,
    'circle': faCircle,
    'circle-dot': faCircleDot,
    'signs-post': faSignsPost,
    'car': faCar,
    'bus': faBus,
    'bicycle': faBicycle,
    'motorcycle': faMotorcycle,
    'truck': faTruck,
    'train': faTrain,
    'plane': faPlane,
    'helicopter': faHelicopter,
    'ship': faShip,
    'taxi': faTaxi,
    'gas-pump': faGasPump,
    'parking': faParking,
    'traffic-light': faTrafficLight,
    'road': faRoad,
    'route': faRoute,
    'truck-medical': faTruckMedical,
    'person-biking': faPersonBiking,
    'person-hiking': faPersonHiking,
    'plug': faPlug,
    'hospital': faHospital,
    'house-medical': faHouseMedical,
    'pills': faPills,
    'heart-pulse': faHeartPulse,
    'syringe': faSyringe,
    'stethoscope': faStethoscope,
    'wheelchair': faWheelchair,
    'cross': faCross,
    'kit-medical': faKitMedical,
    'person-cane': faPersonCane,
    'school': faSchool,
    'building-columns': faBuildingColumns,
    'graduation-cap': faGraduationCap,
    'chalkboard-user': faChalkboardUser,
    'microscope': faMicroscope,
    'book-open': faBookOpen,
    'pencil': faPencil,
    'book': faBook,
    'utensils': faUtensils,
    'mug-hot': faMugHot,
    'store': faStore,
    'cart-shopping': faCartShopping,
    'bed': faBed,
    'landmark': faLandmark,
    'envelope': faEnvelope,
    'scissors': faScissors,
    'wrench': faWrench,
    'city': faCity,
    'building': faBuilding,
    'industry': faIndustry,
    'warehouse': faWarehouse,
    'shield-halved': faShieldHalved,
    'fire-extinguisher': faFireExtinguisher,
    'life-ring': faLifeRing,
    'triangle-exclamation': faTriangleExclamation,
    'bell': faBell,
    'person-shelter': faPersonShelter,
    'vault': faVault,
    'futbol': faFutbol,
    'dumbbell': faDumbbell,
    'masks-theater': faMasksTheater,
    'camera': faCamera,
    'person-swimming': faPersonSwimming,
    'mountain-sun': faMountainSun,
    'tree': faTree,
    'music': faMusic,
    'panorama': faPanorama,
    'church': faChurch,
    'place-of-worship': faPlaceOfWorship,
    'chess-rook': faChessRook,
    'bolt': faBolt,
    'droplet': faDroplet,
    'tower-broadcast': faTowerBroadcast,
    'solar-panel': faSolarPanel,
    'recycle': faRecycle,
    'water': faWater,
    'leaf': faLeaf,
    'fire': faFire,
    'satellite': faSatellite,
    'circle-exclamation': faCircleExclamation,
    'circle-info': faCircleInfo,
    'circle-question': faCircleQuestion,
    'ban': faBan,
    'circle-check': faCircleCheck,
    'arrow-up': faArrowUp,
    'arrow-right': faArrowRight,
};

// ─────────────────────────────────────────────────────────────────────────────
// Gerador de pino SVG (teardrop com ícone FA branco dentro)
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Gera um SVG de pino estilo teardrop com o ícone FontAwesome dentro.
 * @param bgColor  Cor de fundo do pino (hex ou qualquer valor CSS)
 * @param iconDef  IconDefinition do FontAwesome (opcional — pino vazio se ausente)
 * @param size     Tamanho do viewBox do pino (padrão 40×50)
 */
function _makePinSvg(bgColor: string, iconDef?: IconDefinition): string {
    const pinPath =
        'M20 2 C10.1 2 2 10.1 2 20 C2 32 20 50 20 50 C20 50 38 32 38 20 C38 10.1 29.9 2 20 2 Z';

    let iconSvg = '';
    if (iconDef) {
        const raw = iconDef.icon;
        const w = raw[0] as number;
        const h = raw[1] as number;
        const pathData = Array.isArray(raw[4]) ? (raw[4] as string[]).join(' ') : (raw[4] as string);
        // Ícone posicionado centralizado no topo do pino (área ciruclar ~y 2-38)
        iconSvg = `<svg x="8" y="6" width="24" height="24" viewBox="0 0 ${w} ${h}">
      <path d="${pathData}" fill="white"/>
    </svg>`;
    }

    return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 40 50" width="40" height="50">
  <path d="${pinPath}" fill="${bgColor}" stroke="rgba(0,0,0,0.3)" stroke-width="1.5"/>
  ${iconSvg}
</svg>`;
}

/**
 * Gera uma data URL SVG de pino para uso em ol/style/Icon.
 * @param iconKey  Chave kebab-case do catálogo (ex: 'location-dot')
 * @param color    Cor do pino
 * @param _size    Ignorado internamente — o tamanho é controlado pelo scale do ol/style/Icon
 */
export function makePointMarkerUrl(iconKey: string, color: string, _size = 28): string {
    const iconDef = POINT_ICONS[iconKey];
    const svg = _makePinSvg(color, iconDef);
    return 'data:image/svg+xml;charset=utf-8,' + encodeURIComponent(svg);
}

/**
 * Gera data URL de pino simples (sem ícone) para fontes externas (png/svg upload)
 * onde a imagem é carregada separadamente.
 */
export function makePlainPinUrl(color: string): string {
    const svg = _makePinSvg(color, undefined);
    return 'data:image/svg+xml;charset=utf-8,' + encodeURIComponent(svg);
}

File: core/dialog/point_icons.py
# -*- coding: utf-8 -*-
"""
Catalogo de icones FontAwesome Free para camadas de ponto.
Cada entrada: (fa_kebab_key, rotulo_ui)

fa_kebab_key -> camelCase import no TypeScript: 'location-dot' -> faLocationDot
Todos os icones listados existem no @fortawesome/free-solid-svg-icons.
"""
from collections import OrderedDict

POINT_ICON_CATEGORIES = OrderedDict([
    ('Marcadores', [
        ('location-dot',        'Localizacao'),
        ('map-pin',             'Alfinete'),
        ('map-marker-alt',      'Marcador de Mapa'),
        ('star',                'Estrela'),
        ('flag',                'Bandeira'),
        ('crosshairs',          'Mira / Referencia'),
        ('compass',             'Bussola'),
        ('circle',              'Circulo / Ponto'),
        ('circle-dot',          'Ponto Central'),
        ('signs-post',          'Marco / Placa'),
    ]),
    ('Transporte', [
        ('car',                 'Carro'),
        ('bus',                 'Onibus / Terminal'),
        ('bicycle',             'Bicicleta / Ciclovia'),
        ('motorcycle',          'Motocicleta'),
        ('truck',               'Caminhao / Carga'),
        ('train',               'Trem / Metro'),
        ('plane',               'Aeroporto / Aviao'),
        ('helicopter',          'Helicoptero'),
        ('ship',                'Porto / Embarcacao'),
        ('taxi',                'Taxi'),
        ('gas-pump',            'Posto de Combustivel'),
        ('parking',             'Estacionamento'),
        ('traffic-light',       'Semaforo'),
        ('road',                'Rodovia / Via'),
        ('route',               'Rota / Percurso'),
        ('truck-medical',       'Ambulancia'),
        ('person-biking',       'Ciclismo'),
        ('person-hiking',       'Trilha / Caminhada'),
        ('plug',                'Recarga EV'),
    ]),
    ('Saude', [
        ('hospital',            'Hospital'),
        ('house-medical',       'Clinica / UBS / ACS'),
        ('pills',               'Farmacia / Drogaria'),
        ('heart-pulse',         'UPA / Pronto-Socorro'),
        ('syringe',             'Posto de Vacinacao'),
        ('stethoscope',         'Consultorio Medico'),
        ('wheelchair',          'Acessibilidade'),
        ('cross',               'Cruz Vermelha / SAMU'),
        ('kit-medical',         'Primeiros Socorros'),
        ('person-cane',         'CRAS / Assistencia Social'),
    ]),
    ('Educacao', [
        ('school',              'Escola / CMEI'),
        ('building-columns',    'Universidade / Faculdade'),
        ('book',                'Biblioteca'),
        ('graduation-cap',      'Graduacao / Pos-Graduacao'),
        ('chalkboard-user',     'Sala de Aula / Curso'),
        ('microscope',          'Laboratorio / Pesquisa'),
        ('book-open',           'Leitura / Centro Cultural'),
        ('pencil',              'Sala de Recursos / Reforco'),
    ]),
    ('Comercio e Servicos', [
        ('utensils',            'Restaurante / Lanchonete'),
        ('mug-hot',             'Cafe / Padaria'),
        ('store',               'Comercio / Loja'),
        ('cart-shopping',       'Supermercado'),
        ('bed',                 'Hotel / Pousada'),
        ('landmark',            'Banco / Agencia Financeira'),
        ('envelope',            'Correios / Agencia Postal'),
        ('scissors',            'Barbearia / Salao de Beleza'),
        ('wrench',              'Mecanica / Oficina'),
        ('city',                'Centro Comercial / Shopping'),
        ('building',            'Edificio / Sede'),
        ('industry',            'Industria / Fabrica'),
        ('warehouse',           'Deposito / Armazem'),
    ]),
    ('Seguranca e Emergencia', [
        ('shield-halved',       'Delegacia / Policia'),
        ('fire-extinguisher',   'Corpo de Bombeiros'),
        ('life-ring',           'Guarda-Vidas'),
        ('triangle-exclamation', 'Aviso / Area de Risco'),
        ('bell',                'Alarme / Sirene'),
        ('person-shelter',      'Abrigo / Defesa Civil'),
        ('vault',               'Tesouraria / Custodia'),
    ]),
    ('Lazer e Cultura', [
        ('futbol',              'Campo / Estadio de Futebol'),
        ('dumbbell',            'Academia / Ginasio'),
        ('masks-theater',       'Teatro / Cinema / Arte'),
        ('camera',              'Ponto Turistico / Mirante'),
        ('person-swimming',     'Piscina / Praia / Clube'),
        ('mountain-sun',        'Parque / Area Natural'),
        ('tree',                'Praca / Arborizacao'),
        ('music',               'Show / Evento / Arena'),
        ('panorama',            'Vista Panoramica'),
    ]),
    ('Religiao e Patrimonio', [
        ('church',              'Igreja / Catedral'),
        ('place-of-worship',    'Mesquita / Templo / Sinagoga'),
        ('chess-rook',          'Fortaleza / Castelo / Patrimonio'),
    ]),
    ('Infraestrutura e Meio Ambiente', [
        ('bolt',                'Subestacao / Energia Eletrica'),
        ('droplet',             'ETA / Captacao / Fonte'),
        ('tower-broadcast',     'Antena / Telecomunicacoes'),
        ('solar-panel',         'Usina Solar'),
        ('recycle',             'Reciclagem / Coleta Seletiva'),
        ('water',               'Manancial / Rio / Lago'),
        ('leaf',                'Area Verde / Ambiental'),
        ('fire',                'Queimada / Risco de Incendio'),
        ('industry',            'Fabrica / Industria'),
    ]),
    ('Sinais e Avisos', [
        ('circle-exclamation',  'Atencao / Alerta'),
        ('circle-info',         'Informacao / Orientacao'),
        ('circle-question',     'Consulta Publica / Duvida'),
        ('ban',                 'Proibicao / Restricao'),
        ('circle-check',        'Liberado / Aprovado'),
        ('arrow-up',            'Sentido Norte / Direcao'),
        ('arrow-right',         'Direcao Leste / Continuacao'),
    ]),
])

# ------------------------------------------------------------------
# Estilos padrao por geometria
# ------------------------------------------------------------------
DEFAULT_POINT_STYLE = {
    'source':     'icon',
    'icon_key':   'location-dot',
    'color':      '#000000',
    'size':       28,
    'scale_zoom': False,
    'zoom_base':  14,
    'min_size':   16,
    'max_size':   40,
    'opacity':    1.0,
    'file_data':  None,
    'file_name':  '',
}

DEFAULT_LINE_STYLE = {
    'color':   '#000000',
    'width':   2,
    'dash':    'solid',
    'opacity': 1.0,
}

DEFAULT_POLYGON_STYLE = {
    'fill_color':     '#000000',
    'fill_opacity':   0.3,
    'stroke_color':   '#000000',
    'stroke_width':   2,
    'stroke_dash':    'solid',
    'stroke_opacity': 1.0,
}


def fa_key_to_ts(key: str) -> str:
    """Converte kebab-case para camelCase com prefixo 'fa'.
    Ex.: 'location-dot' -> 'faLocationDot'
    """
    return 'fa' + ''.join(w.capitalize() for w in key.split('-'))


def all_fa_ts_names() -> list:
    """Retorna todos os nomes camelCase unicos do catalogo."""
    names = set()
    for entries in POINT_ICON_CATEGORIES.values():
        for key, _ in entries:
            names.add(fa_key_to_ts(key))
    return sorted(names)
